torch/cuda/graphs.py

```diff
@@ -0,0 +1,89 @@
+import ctypes
+import gc
+import torch
+
+from ._utils import _dummy_type
+
+
+if not hasattr(torch._C, '_CudaStreamBase'):
+    # Define dummy base classes
+    torch._C.__dict__['CUDAGraph'] = _dummy_type('CUDAGraph')
+    torch._C.__dict__['graph_pool_handle'] = _dummy_type('graph_pool_handle')
+
+from torch._C import (
+    CUDAGraph,
+    graph_pool_handle
+)
+
+
+class graph(object):
+    r"""Context-manager that captures CUDA work into a :class:`torch.cuda.CUDAGraph`
+    object for later replay.
+
+    See :ref:`CUDA Graphs <cuda-graph-semantics>` for a general introduction,
+    detailed use, and constraints.
+
+    Arguments:
+        graph (torch.cuda.CUDAGraph): Graph object used for capture.
+        pool: Opaque token (returned by a call to :func:`~torch.cuda.graph_pool_handle()` or :meth:`other_Graph_instance.pool()<torch.cuda.CUDAGraph.pool>`) hinting this graph's capture may share memory from the specified pool.
+        stream (torch.cuda.Stream, optional): If supplied, will be set as the current ambient stream in the context.  If not supplied, ``Graph`` sets its own internal side stream as the ambient stream in the context.
+
+    .. note::
+        For effective memory sharing, if you pass a ``pool`` used by a previous capture and the previous capture
+        used an explicit ``stream`` argument, you should pass the same ``stream`` argument to this capture.
+
+    """
+    # default_capture_stream = torch.cuda.Stream()
+
+    def __init__(self,
+                 graph,
+                 pool=None,
+                 stream=None):
+        self.capture_stream = torch.cuda.Stream() if stream is None else stream
+        self.stream_ctx = torch.cuda.stream
+        self.graph = graph
+
+    def __enter__(self):
+        # Free as much memory as we can for the graph
+        torch.cuda.synchronize()
+        gc.collect()
+        torch.cuda.empty_cache()
```

Comment:
Nyeh, you sure about this?

Reply:
This has been the way to go in local experiments: free as much memory as we can for the graph pool, then let the graph build up the memory it needs during capture, including by calling cudaMalloc. The native allocator's cudaMallocs are allowed during capture, and will behave well: create some new addresses for the graph to bake in, then become no-ops (ie, they won't allocate more memory) in the replay.